app/controllers/api/v2/split_registries_controller.rb

```diff
@@ -2,6 +2,17 @@ class Api::V2::SplitRegistriesController < UnauthenticatedApiController
   include CorsSupport
 
   def show
-    @split_registry = SplitRegistry.instance
+    build_timestamp = BuildTimestamp.new(build_params)
+    if build_timestamp.valid?
+      @split_registry = SplitRegistry.new(build_timestamp.build_timestamp)
```

Comment:
Seeing this all live makes me want to rethink our approach a bit. The relationship between timestamp and split registry feels a little tenuous. It's not immediately clear from the API why you would be passing in a timestamp, and what kind of output that would lead to.

We had toyed with the idea of using a different name, something like `split_registries_snapshot` to represent the (quasi-)point-in-time nature of the results we're returning.

If we went down that route, we'd keep the original route, add a new one with the above name, and then combine `BuildTimestamp` and `SplitRegistry` model into a new one called `SplitRegistrySnapshot` that would take in the timestamp.

With this approach, there's a clear relationship between the endpoint name and the param it accepts. And it allows us to have a model that encapsulates all of this.

What do you think?

cc @smudge @samandmoore would love your thoughts too!

Reply:
Interesting. This makes me think that the contract between apps isn't entirely clear -- if it isn't too late, I'd be inclined to propose renaming `timestamp` to `as_of`, so that the semantics of asking for a "split registry" make sense when you tag on "?as_of=XXYYZZ" -- adding "snapshot" to the name then feels a little redundant in that world. (but, does feel necessary when it isn't clear what "timestamp" means or does)

Reply:
Yeah, `as_of` would make it clear within the code, but the route would still be ambiguous:

`/api/v2/split_registries/2019-12-13T00:00:00`

vs

`/api/v2/split_registry_snapshots/2019-12-13T00:00:00`

I suppose if we had documentation, we could give the parameter a name (like `as_of`), but it would be nice if the url was somewhat self-documenting.

Reply:
Just reread your comment. It makes sense if we use a query param, but since this is a required param, we're making it part of the path.

Reply:
Ah, ok -- I guess I should've looked closer at the routes 😅 -- so if we're committed to putting it in the path then we can't name it, and if we can't name it, then yeah that forces design considerations around the resource name.

But it's not exactly a "snapshot," right? It's a lookup of the most recent state prior to the timestamp (as in, not an exact match, like with a true ID lookup). IMO for other query/search-ey endpoints, it's fine to have a required param that lives in the technically optional query param list -- you just return a 5XX error if it's missing.

e.g. `/search?q=test+123`

I get that this isn't actually a _search_ endpoint, but it's also not pulling up a "snapshot" with an exact match on the ID param -- it's an "as of" query.

Another interesting thing is how this is gonna appear in the logs/APM/etc -- typically the set of possible IDs is somewhat bounded, but in this case the "ID" is a timestamp, which means it'll be like making a bunch of web requests to `resource/:id` with ever-increasing ids that mostly never overlap in the same way that DB ids would. So you'd better hope that the APM dashboard is good at pattern matching and turning it into a wildcard! I'd assume it would, but I've also never tried it with a formatted timestamp before... 🤔

Reply:
Also, sorry for the wall of text, I'm just trying to get all of my thoughts down in one place! If we end up going with a resource name change -- i.e. `/api/v2/split_registry_snapshots/2019-12-13T00:00:00` -- my only thought would be to double check and make sure our APM solution can handle timestamps as resource IDs.

Reply:
> my only thought would be to double check and make sure our APM solution can handle timestamps as resource IDs.

We already do this [for the app_visitor_configs route](https://github.com/Betterment/test_track/blob/master/config/routes.rb#L29), so I'm not worried about adding another similar case. It looks good in the APM too.

> IMO for other query/search-ey endpoints, it's fine to have a required param that lives in the technically optional query param list -- you just return a 5XX error if it's missing.

Yeah, I don't have a huge problem with required query params, though I think search might be the only great example of this. In this case, the latest refactor shows the value of treating this like a standard REST resource with an identifier parameter. It provides a single encapsulated model, and there's nice alignment between the route and model name.

I'm with you that it's stretching the notion of "snapshot". I'm open to other names if you have any 😄.

My defense of "snapshot" is that it serves that from the client's perspective. Its a (mostly) consistent result that gives the client a point in time representation of the split registry. Splits weights will change and we will ship newly created splits. Clients will ignore newly created splits, and the updated weight... well, that's how TT works. 🤷‍♂

Reply:
`/api/v2/builds/[timestamp]/split_registry` seems like it'd be the most coherent with the mobile route. We did a lot of this debate before settling on the URL structure of the mobile route - we could reopen it, but we def talked about optional vs required query params and naming of things, etc.

Reply:
fwiw it looks like we tag our logs with a timestamp already (not, i don't think, related to the app_visitor_configs route) 
![image](https://user-images.githubusercontent.com/14004878/70934768-a8170a00-200c-11ea-94ab-dae10652a7a4.png)

My 2¢ is that I like how this makes the route quite _resource-y_. It really is not a search, nor is this a *true* snapshot, but I think it comes close enough by returning a mostly reproducible resource for a given id (timestamp). That being said doing something like `/api/v2/builds/[timestamp]/split_registry` seems nice especially if it conforms with the mobile standard that has been set.